Tribler/community/popularity/pubsub.py

```diff
@@ -0,0 +1,239 @@
+
+        self.register_task("start_publishing", LoopingCall(start_publishing)).start(PUBLISH_INTERVAL, False)
+
+    @inlineCallbacks
+    def unload(self):
+        self.request_cache.clear()
+        self.cancel_pending_task("start_publishing")
+        yield super(PubSubCommunity, self).unload()
+
+    def subscribe_peers(self):
+        """
+        Subscribes to the connected peers. First, the peers are sorted based on the trust score on descending order and
+        content subscribe request is sent to the top peers.
+        This method is called periodically through refresh_peer_list() in start_publishing() loop so it can fill up for
+        the disconnected peers by connecting to new peers.
+        Note that, existing publisher peers are not disconnected even if we find new peers with higher trust score but
+        only fill up the remaining publisher slots with new top peers.
+        """
+        num_publishers = len(self.publishers)
+        num_peers = len(self.get_peers())
+        # If we have some free publisher slots and there are peers available
+        if num_publishers < MAX_PUBLISHERS and num_publishers < num_peers:
+            available_publishers = [peer for peer in self.get_peers() if peer not in self.publishers]
+            sorted_peers = sorted(available_publishers,
+                                  key=lambda _peer: self.trustchain.get_trust(_peer) if self.trustchain else 1,
+                                  reverse=True)
+            for peer in sorted_peers[: MAX_PUBLISHERS - num_publishers]:
+                self.subscribe(peer, subscribe=True)
+
+    def refresh_peer_list(self):
+        """
+        Updates the publishers and subscribers list by filtering out the disconnected peers. It also calls subscribe
+        peers to replenish the available publisher slots if necessary.
+        """
+        peers = self.get_peers()
+        self.publishers = set([peer for peer in self.publishers if peer in peers])
+        self.subscribers = set([peer for peer in self.subscribers if peer in peers])
+
+        # subscribe peers if necessary
+        self.subscribe_peers()
+
+    def unsubscribe_peers(self):
+        """
+        Unsubscribes from the existing publishers by sending content subscribe request with subscribe=False. It then
+        clears up its publishers list.
+        - Called at community unload.
+        """
+        for peer in copy(self.publishers):
+            self.subscribe(peer, subscribe=False)
+        self.publishers.clear()
+
+    def subscribe(self, peer, subscribe=True):
+        """
+        Method to send content subscribe/unsubscribe message. This message is sent to each individual publisher peer we
+        want to subscribe/unsubscribe.
+        """
+        cache = self.request_cache.add(ContentRequest(self.request_cache, MSG_SUBSCRIBE, None))
+        # Add or remove the publisher peer
+        if subscribe:
+            self.publishers.add(peer)
```

Comment:
Despite you're using a `set` to store the `publishers`, you should still only add them on receiving successful `ContentSubscription` message. Otherwise, the confirmation message is useless, and you become vulnerable to packet loss.

Please, make `publishers` only be added on successful confirmation message.
(And BTW, do subscriptions time out?)

Reply:
Good point! will make the change.
Subscriptions do not time out. However, there are periodically refreshed to fill up for disconnected peers.